### PR TITLE
[query] ParseTime function supports 95%+ of `from / until` time formats

### PR DESCRIPTION
**Pull request by teddywahle**
**What this PR does / why we need it**:

TLDR: We need a lot of code to support the long tail of obscure Graphite time inputs.

Graphite has a time format that it refers to as `from / until` in many places in its [documentation](https://graphite.readthedocs.io/en/latest/render_api.html#from-until).

`from/until` supports an INSANELY varied formatting structure:
Screen Shot 2020-09-09 at 9 15 17 PM

The difficult part here is that the actual Graphite source code does not support all of those examples. However, it does support other examples that were not explained anywhere in the documentation. Because of this, in this PR, I basically copied the logic of Graphite's Python  source code, and I translated it into Go.

You can view that [source code here](https://github.com/graphite-project/graphite-web/blob/master/webapp/graphite/render/attime.py).

I also basically copied all of their test cases. You will see a few that are missing. The only ones I excluded were tests for formatting styles that are not explained anywhere in Graphite's documentation. 

Here are links to [Graphite's test cases](https://github.com/graphite-project/graphite-web/blob/master/webapp/tests/test_attime.py).

An example of a test case that I omitted is the test of `ParseReference` that takes `midnight_tomorrow` as an input. This format of `timeString_relativeDayString` is explained nowhere in the docs. Also, Graphite claims that these time schemas are compatible with [at(1)](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/at.html) format, but they are not. I don't know how a user would ever think to format their input like this.

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/query/graphite/graphite/timespec.go [hunk 1]
```diff
@@ -94,6 +123,7 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
 		return now.Add(-1 * time.Duration(timePast) * period), nil
 	}
 
+
```
Comment (teddywahle) [inline src/query/graphite/graphite/timespec.go:126]:
```suggestion
```

File: src/query/graphite/graphite/timespec.go [hunk 2]
```diff
@@ -108,9 +137,146 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+	if strings.Contains(s, "+") {
+		stringSlice := strings.Split(s, "+")
+		ref, offset = stringSlice[0], stringSlice[1]
+		offset = "+" + offset
+	} else if strings.Contains(s, "-") {
+		stringSlice := strings.Split(s, "-")
+		ref, offset = stringSlice[0], stringSlice[1]
+		offset = "-" + offset
+	}
+	err = nil
```
Comment (robskillington) [inline src/query/graphite/graphite/timespec.go:150]: No need to make err = nil, it will be shadowed/rewritten below, it's not required to reset it here.

File: src/query/graphite/graphite/timespec.go [hunk 3]
```diff
@@ -108,9 +137,146 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+	parsedReference, err := ParseTimeReference(ref, now)
```
Comment (robskillington) [inline src/query/graphite/graphite/timespec.go:151]: Need to check err here before proceeding to next line?

Maybe this could be rewritten:
```go
if parsedReference, err := ParseTimeReference(ref, now); err == nil {
  if parsedOffset, err := ParseOffset(offset); err == nil {
    return parsedReference.Add(parsedOffset), nil
  }
}
```

File: src/query/graphite/graphite/timespec.go [hunk 4]
```diff
@@ -108,9 +137,146 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+	}
+
 	return now, err
 }
 
+
+func ParseTimeReference(ref string, now time.Time) (time.Time, error) {
+	hour := now.Hour()
+	minute := now.Minute()
+	refDate := time.Time{}
```
Comment (robskillington) [inline src/query/graphite/graphite/timespec.go:164]: Put these into a var block (so you don't need to explicitly declare refDate as zero value):
```go
var (
  hour = now.Hour()
  minute = now.Minute()
  refDate time.Time
)
```

File: src/query/graphite/graphite/timespec.go [hunk 5]
```diff
@@ -108,9 +137,146 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+
+	// Time of Day Reference
+	i := strings.Index(ref,":")
+	if 0 < i && i < 3 {
+		newHour, err := strconv.ParseInt(ref[:i], 10, 0)
+		if err != nil {
+			return time.Time{}, err
+		}
+		hour = int(newHour)
+		newMinute, err := strconv.ParseInt(ref[i+1:i+3], 10, 32)
```
Comment (robskillington) [inline src/query/graphite/graphite/timespec.go:188]: Should we be bounds check i+FOO?

File: src/query/graphite/graphite/timespec.go [hunk 6]
```diff
@@ -108,9 +137,146 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+		if err != nil {
+			return time.Time{}, err
+		}
+		minute = int(newMinute)
+		ref = ref[i+3:]
```
Comment (robskillington) [inline src/query/graphite/graphite/timespec.go:193]: Should we be bounds check i+FOO?

File: src/query/graphite/graphite/timespec.go [hunk 7]
```diff
@@ -108,9 +137,146 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+
+		if len(ref) >= 2 {
+			if ref[:2] == "am" {
+				ref = ref[2:]
+			} else if  ref[:2] == "pm" {
+				hour = (hour + 12) % 24
+				ref = ref[2:]
```
Comment (robskillington) [inline src/query/graphite/graphite/timespec.go:200]: Should we be bounds check i+FOO?

File: src/query/graphite/graphite/timespec.go [hunk 8]
```diff
@@ -108,9 +137,146 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+	// Xam or XXam
+	i = strings.Index(ref,"am")
+	if 0 < i && i < 3 {
+		newHour, err := strconv.ParseInt(ref[:i], 10, 32)
+		if err != nil {
+			return time.Time{}, err
+		}
+		hour = int(newHour)
+		minute = 0
+		ref = ref[i+2:]
```
Comment (robskillington) [inline src/query/graphite/graphite/timespec.go:214]: Should we be bounds check i+FOO?

File: src/query/graphite/graphite/timespec.go [hunk 9]
```diff
@@ -108,9 +137,146 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+		expectedDay := weekdays[ref]
+		today := int(refDate.Weekday())
+		dayOffset := today - expectedDay
+		if dayOffset < 0 {
+			dayOffset += 7
+		}
+		offSetDuration := time.Duration(dayOffset)
+		refDate = refDate.Add(time.Hour * 24 * -1 * offSetDuration)
+	} else if ref != "" {
+		return time.Time{}, errors.NewInvalidParamsError(fmt.Errorf("unkown day reference %s", rawRef))
```
Comment (robskillington) [inline src/query/graphite/graphite/timespec.go:273]: "unkown" -> "unknown"

File: src/query/graphite/graphite/timespec_test.go [hunk 10]
```diff
@@ -104,3 +129,69 @@ func TestAbsoluteOffset(t *testing.T) {
...
+		"10",
+		"month",
+	}
+
+	for _, test := range tests {
+		parsed, err := ParseOffset(test)
+		assert.Error(t, err)
+		assert.Equal(t, time.Duration(0), parsed)
+	}
+}
```
Comment (robskillington) [inline src/query/graphite/graphite/timespec_test.go:197]: Should have empty endline at end of file.

File: src/query/graphite/graphite/timespec.go [hunk 11]
```diff
@@ -108,9 +137,146 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+		ref, offset = stringSlice[0], stringSlice[1]
+		offset = "-" + offset
+	}
+	err = nil
+	parsedReference, err := ParseTimeReference(ref, now)
+	parsedOffset, err := ParseOffset(offset)
+	if err == nil {
+		return parsedReference.Add(parsedOffset), err
+	}
+
```
Comment (teddywahle) [inline src/query/graphite/graphite/timespec.go:156]:
```suggestion
	parsedReference, err := ParseTimeReference(ref, now)
if err == nil {
  parsedOffset, err := ParseOffset(offset)
  if err == nil {
    return parsedReference.Add(parsedOffset), nil
  }
}

```

File: src/query/graphite/graphite/timespec.go (hunk 3 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go:150]: great idea

File: src/query/graphite/graphite/timespec.go [hunk 12]
```diff
@@ -108,9 +137,147 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+}
+
 	return now, err
 }
 
+
+func ParseTimeReference(ref string, now time.Time) (time.Time, error) {
+	hour := now.Hour()
+	minute := now.Minute()
+	refDate := time.Time{}
```
Comment (teddywahle) [inline src/query/graphite/graphite/timespec.go:165]:
```suggestion
	var (
  hour = now.Hour()
  minute = now.Minute()
  refDate time.Time
)
```

File: src/query/graphite/graphite/timespec.go [hunk 13]
```diff
@@ -108,9 +137,149 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+		expectedDay := weekdays[ref]
+		today := int(refDate.Weekday())
+		dayOffset := today - expectedDay
+		if dayOffset < 0 {
+			dayOffset += 7
+		}
+		offSetDuration := time.Duration(dayOffset)
+		refDate = refDate.Add(time.Hour * 24 * -1 * offSetDuration)
+	} else if ref != "" {
+		return time.Time{}, errors.NewInvalidParamsError(fmt.Errorf("unkown day reference %s", rawRef))
```
Comment (teddywahle) [inline src/query/graphite/graphite/timespec.go:276]:
```suggestion
		return time.Time{}, errors.NewInvalidParamsError(fmt.Errorf("unknown day reference %s", rawRef))
```

File: src/query/graphite/graphite/timespec_test.go (hunk 10 above)
Comment (teddywahle) [inline src/query/graphite/graphite/timespec_test.go:197]:
```suggestion
}

```

File: src/query/graphite/graphite/timespec.go (hunk 5 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go:190]: Good call. I added a bounds check and a new test to make sure the bounds check is working.

File: src/query/graphite/graphite/timespec.go (hunk 2 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go]: Got it

Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go]: OK, good to know. Thanks.

File: src/query/graphite/graphite/timespec.go (hunk 9 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go]: Done

File: src/query/graphite/graphite/timespec_test.go (hunk 10 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec_test.go]: Done

File: src/query/graphite/graphite/timespec_test.go [hunk 14]
```diff
@@ -104,3 +129,70 @@ func TestAbsoluteOffset(t *testing.T) {
...
```
Comment (teddywahle) [inline src/query/graphite/graphite/timespec_test.go:198]:
```suggestion
}

```

File: src/query/graphite/graphite/timespec_test.go (hunk 10 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec_test.go]: hmm, any time i add this, it gets removed by `go fmt`

File: src/query/graphite/graphite/timespec.go [hunk 15]
```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+		parsedOffset, err := ParseOffset(offset)
+		if err == nil {
+			return parsedReference.Add(parsedOffset), nil
+		}
+	}
+
 	return now, err
 }
 
+func ParseTimeReference(ref string, now time.Time) (time.Time, error) {
```
Comment (robskillington) [inline src/query/graphite/graphite/timespec.go:161]: Need to add comment for exported symbols.

File: src/query/graphite/graphite/timespec.go [hunk 16]
```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+		if err != nil {
+			return time.Time{}, err
+		}
+		hour = int(newHour)
+		if len(ref) >= i+3 {
+			newMinute, err := strconv.ParseInt(ref[i+1:i+3], 10, 32)
+			if err != nil {
+				return time.Time{}, err
+			}
+			minute = int(newMinute)
```
Comment (arnikola) [inline src/query/graphite/graphite/timespec.go:195]: Should this make sure it's not over 60 or is that allowed?

File: src/query/graphite/graphite/timespec.go [hunk 17]
```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+			ref = ref[i+3:]
+		}
+
+		if len(ref) >= 2 {
+			if ref[:2] == "am" {
+				ref = ref[2:]
+			} else if ref[:2] == "pm" {
+				hour = (hour + 12) % 24
```
Comment (arnikola) [inline src/query/graphite/graphite/timespec.go:203]: Is this valid for e.g. `23:00PM` or should that be a parse error

File: src/query/graphite/graphite/timespec.go [hunk 18]
```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+	}
+
+	// Xam or XXam
+	i = strings.Index(ref, "am")
+	if 0 < i && i < 3 {
+		newHour, err := strconv.ParseInt(ref[:i], 10, 32)
+		if err != nil {
+			return time.Time{}, err
+		}
+		hour = int(newHour)
```
Comment (arnikola) [inline src/query/graphite/graphite/timespec.go:216]: Wouldn't this parse something like `10:00pm6am`?

File: src/query/graphite/graphite/timespec.go [hunk 19]
```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+				ref = ref[2:]
+			} else if ref[:2] == "pm" {
+				hour = (hour + 12) % 24
+				ref = ref[2:]
+			}
+		}
+	}
+
+	// Xam or XXam
+	i = strings.Index(ref, "am")
```
Comment (arnikola) [inline src/query/graphite/graphite/timespec.go:210]: These should probably be regexing aginst the initial `ref` right?

File: src/query/graphite/graphite/timespec.go [hunk 20]
```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+		if err != nil {
+			return time.Time{}, err
+		}
+		hour = int(newHour)
+		minute = 0
+		ref = ref[i+2:]
+	}
+
+	// Xpm or XXpm
+	i = strings.Index(ref, "pm")
```
Comment (arnikola) [inline src/query/graphite/graphite/timespec.go:222]: These should probably be regexing aginst the initial `ref` right?

File: src/query/graphite/graphite/timespec.go [hunk 21]
```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+	if 0 < i && i < 3 {
+		newHour, err := strconv.ParseInt(ref[:i], 10, 32)
+		if err != nil {
+			return time.Time{}, err
+		}
+		hour = int((newHour + 12) % 24)
```
Comment (arnikola) [inline src/query/graphite/graphite/timespec.go:228]: Is it correct to parse `99pm` or better to fail?

File: src/query/graphite/graphite/timespec.go [hunk 22]
```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+		ref = ref[8:]
+	} else if strings.HasPrefix(ref, "teatime") {
+		hour, minute = 16, 0
+		ref = ref[7:]
+	}
+
+	refDate = time.Date(now.Year(), now.Month(), now.Day(), hour, minute, 0, 0, now.Location())
+
+	// Day reference
+	if ref == "yesterday" {
```
Comment (arnikola) [inline src/query/graphite/graphite/timespec.go:247]: Should these match against the initial `ref` before it's been shortened?

File: src/query/graphite/graphite/timespec.go [hunk 23]
```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
 
+	ref, offset := s, ""
+	if strings.Contains(s, "+") {
+		stringSlice := strings.Split(s, "+")
+		ref, offset = stringSlice[0], stringSlice[1]
+		offset = "+" + offset
+	} else if strings.Contains(s, "-") {
+		stringSlice := strings.Split(s, "-")
+		ref, offset = stringSlice[0], stringSlice[1]
+		offset = "-" + offset
```
Comment (teddywahle) [inline src/query/graphite/graphite/timespec.go:148]:
```suggestion
		stringSlice := strings.Split(s, "-")
		if len(stringSlice) == 2 {
	               ref, offset = stringSlice[0], stringSlice[1]
		      offset = "-" + offset
		}
```

File: src/query/graphite/graphite/timespec.go [hunk 24]
```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
 		return time.Unix(n, 0).UTC(), nil
 	}
 
+	ref, offset := s, ""
+	if strings.Contains(s, "+") {
+		stringSlice := strings.Split(s, "+")
+		ref, offset = stringSlice[0], stringSlice[1]
+		offset = "+" + offset
```
Comment (teddywahle) [inline src/query/graphite/graphite/timespec.go:144]:
```suggestion
	if strings.Contains(s, "+") {
		stringSlice := strings.Split(s, "+")
		if len(stringSlice) == 2 {
		     ref, offset = stringSlice[0], stringSlice[1]
		     offset = "+" + offset
		}
```

File: src/query/graphite/graphite/timespec.go [hunk 25]
```diff
@@ -108,9 +137,152 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+
+	if ref == "" || ref == "now" {
+		return now, nil
+	}
+
+	// check if the time ref fits an absolute time format
+	for _, format := range formats {
+		t, err := time.Parse(format, ref)
+		if err == nil {
+			return time.Date(t.Year(), t.Month(), t.Day(), hour, minute, 0, 0, now.Location()), nil
```
Comment (teddywahle) [inline src/query/graphite/graphite/timespec.go:180]:
```suggestion
			return t nil
```

File: src/query/graphite/graphite/timespec.go [hunk 26]
```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
```
[the remark this answers is not in the document]
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go:180]: yeah, that's much cleaner.

File: src/query/graphite/graphite/timespec.go (hunk 21 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go:232]: Good call. Added a check and an error test case for this.

File: src/query/graphite/graphite/timespec.go (hunk 16 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go:214]: Good call. I will throw an error and add a test case for this.

File: src/query/graphite/graphite/timespec.go (hunk 17 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go:222]: Good call. I will throw an error and add a test case for this.

File: src/query/graphite/graphite/timespec.go (hunk 19 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go:229]: Good call. I changed it to a regex.

File: src/query/graphite/graphite/timespec.go (hunk 20 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go:241]: Good call. I changed it to a regex.

File: src/query/graphite/graphite/timespec.go (hunk 22 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go:269]: Yes. Fixed.

File: src/query/graphite/graphite/timespec.go (hunk 18 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go:240]: Good call. Added a check and an error test case for this.

File: src/query/graphite/graphite/timespec.go (hunk 22 above)
Comment (teddywahle) [reply src/query/graphite/graphite/timespec.go]: No, because this function needs to be able to parse "noon yesterday" which will actually be "noonyesterday" when passed in because whitespace gets stripped. Parsing the unshortened ref could break this.

File: src/query/graphite/graphite/timespec.go [hunk 27]
```diff
@@ -108,9 +152,153 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
 		return time.Unix(n, 0).UTC(), nil
 	}
 
+	s = strings.Replace(strings.Replace(strings.ToLower(s), ",", "", -1), " ", "", -1)
+	ref, offset := s, ""
+	if strings.Contains(s, "+") {
+		stringSlice := strings.Split(s, "+")
+		if len(stringSlice) == 2 {
```
Comment (arnikola) [inline src/query/graphite/graphite/timespec.go:159]: Should this error if len != 2?

File: src/query/graphite/graphite/timespec.go [hunk 28]
```diff
@@ -108,9 +152,153 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
...
+			ref, offset = stringSlice[0], stringSlice[1]
+			offset = "+" + offset
+		}
+	} else if strings.Contains(s, "-") {
+		stringSlice := strings.Split(s, "-")
+		if len(stringSlice) == 2 {
```
Comment (arnikola) [inline src/query/graphite/graphite/timespec.go:165]: Same here

**Review by arnikola — APPROVED** [review]
Stamped w nits
